SIPS/sip-5.md

```diff
@@ -0,0 +1,176 @@
+
+## Specification
+
+> Formal specifications are written in Typescript.
+
+### Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and
+"OPTIONAL" written in uppercase in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt)
+
+### Definitions
+
+JWTs are composed of three parts: header, payload, and signature. For more information, please check [JSON Web Token (JWT) Request for Comments (RFC)](https://www.rfc-editor.org/rfc/rfc7519).
+
+### Snap Manifest
+
+This SIP specifies permission named `snap_signJwt`. This permission grants a snap the ability to create signed JWTs by introducing an additional signing method.
+
+This specification is specified as follows in `snap.manifest.json` files:
+```typescript
+{    
+    "initialPermissions": {
+        "snap_signJwt": {},
+    }
+}
+```
+
+### Common Types
+
+The below common types are used throughout the specification.
+
+```typescript
+interface JWTHeader {
+  typ: string;
+  alg: string;
+  [params: string]: any;
+}
+
+interface JWTPayload {
+  iss?: string;
+  sub?: string;
+  aud?: string | string[];
+  jti?: string;
+  nbf?: number;
+  exp?: number;
+  iat?: number;
+  [claims: string]: any;
+}
+
+type HashingFunction = (data: string | Uint32Array) => string | Uint32Array;
+```
+
+### JWT Signing Implementation
+
+This implementation is modeled after implementations from libraries [did-jwt](https://github.com/decentralized-identity/did-jwt) and [jose](https://github.com/panva/jose).
+
+```typescript
+import * as u8a from "uint8arrays";
+
+function hexToBytes(s: string): Uint8Array {
+  const input = s.startsWith("0x") ? s.substring(2) : s;
+  return u8a.fromString(input.toLowerCase(), "base16");
+}
+
+function bytesToBase64url(b: Uint8Array): string {
+  return u8a.toString(b, "base64url");
+}
+
+function encodeSection(data: any): string {
+  return encodeBase64url(JSON.stringify(data));
+}
+
+export function encodeBase64url(s: string): string {
+  return bytesToBase64url(u8a.fromString(s));
+}
+
+export async function createJWT(
+  header: Partial<JWTHeader> = {},
+  payload: Partial<JWTPayload> = {},
+  address: string,
+  hashingFunction: HashingFucntion
+): Promise<string> {
+  if (!header.typ) header.typ = "JWT";
+    
+  const encodedPayload = (typeof payload === "string") ? payload : encodeSection(payload);
+  const signingInput: string = [encodeSection(header), encodedPayload].join(
+    "."
+  );
+
+  const hash = hashingFunction(signingInput);
+  const bytes = hexToBytes(hash);
+
+  let signature = /* eth_sign MetaMask RPC method that creates the signature (showing the user header and payload that will be signed) */;
```

Comment:
Do you have any more details on this? What's the actual payload to sign?

Reply:
Yes, sorry for the ambiguity. It is `bytes` from line 110.